### DISCUSSION TITLE
Nutshell: Broken Promises on Intermediary Data

### DISCUSSION THREAD

**Discussion by caseywalker** [nss-evening-cohort-14/evening-client-side#215] [Help Me!] [answered]
# Casey

* Have you searched the other help tickets? Yesh

* Describe your issue:
- [ ] Oh boy, where do I begin? 
- [ ] I'm trying to resolve a Promise.all on intermediary data that contains a dinosaur's firebaseKey, and the assigned staff member. 
- [ ] I have two separate promises. One is returning an array of intermediary data. The second promise, which is the broken promise, takes that response and I want to loop through each of the staff's firebaseKeys and dino's firebaseKeys to return staff.title and dino.species
- [ ] When testing and console.warning, I was getting an array of dino species and staff name. 
- [ ] However, when I chain a .then on domEvents, only one staff and dino is being returned
- [ ] I'm thoroughly impressed if you are still reading at this point, I would have given up a long time ago 

# TO TEST! 

1. npm start
2. sign in plz
3. click 'Staff' icon
4. kindly click 'View Assigned Staff' button
5. inspect the console. 

* List any error messages that appear
- [ ] None. Just not getting the results desired. 

* What have you googled? 
This is about how I was structuring the data, as I feel that may be a portion of the problem
1. Do I want to map? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
2. Let's try a forEach https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach (current data structure)
3. Maybe let's just get the object values? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Object/values

# What have you tried? Tell us everything you have tried.
* I've tried refactoring my promise and chaining .thens in so many different fashions that I don't know where to go at this point. 

* Copy/paste link TO THE FILE on your branch or in your repo 
Second promise below, it is expecting a response from the promise above. 
https://github.com/nss-evening-cohort-14/jurrasic-world-dashboard-jurrasic-world-dashboard/blob/cw-sr-assignments/src/javascripts/helpers/data/assignmentData.js#L17

Dom Events for calling this
* https://github.com/nss-evening-cohort-14/jurrasic-world-dashboard-jurrasic-world-dashboard/blob/cw-sr-assignments/src/javascripts/helpers/events/domEvents.js#L390

**Comment by drteresavasquez** [comment]
I see this and will look at it tomorrow. Another team had a similar issue and I’m going to get them to drop some info too.

**Reply by drteresavasquez** [reply]
Also, @caseywalker have you reviewed the videos? It sounds like your answer may be in there.

**Reply by drteresavasquez** [reply]
https://github.com/nss-evening-cohort-14/evening-client-side/discussions/213

**Comment by drteresavasquez** [comment]
OK. I think I understand your situation. The reason you are only getting one back is because you are only resolving one object. When you tell the promise to resolve, it is almost like a return statement. 

What you could try to do is change your `forEach` to a `map` and return the object to the map. Since it is looping, it will apply that to all of the items that you are looping over and return an array of those promises.

### NOTE: I did not run this so it may need some updates.

Here is the important thing to know. When you are looping through ANY array and using a promise inside, you are creating a new promise that will need to be resolved (exactly like you do in DOM events when you call getAssignmentStaff and you have to .then it), so if you have 50 elements in the array you are looping through, you just created 50+ promises depending on how many elements/promises you have. In your code, you are using `Promise.all` on 2 promises, so you are doubling your promise load.

In the push or map, we are creating an array of Promises, you then would use Promise.all to get the data and then resolve it.

Your code looks like this:
```javascript
const getAssignmentStaff = (object) => new Promise((resolve, reject) => {
  const dinoStaff = object.dino;
  console.warn(dinoStaff);
  dinoStaff.forEach((element) => {
    const staffKey = element.staff_firebaseKey;
    const dinoKey = element.dino_firebaseKey;
    const promiseOne = getSingleDino(dinoKey);
    const promiseTwo = getSingleStaffMember(staffKey);

    Promise.all([promiseOne, promiseTwo]).then(([dinoResponse, StaffResponse]) => resolve({ dino: dinoResponse.species, staff: StaffResponse.name })).catch((error) => reject(error));
  });
});
```

# Option 1: Refactor
```javascript
const getAssignmentStaff = (object) => new Promise((resolve, reject) => {
  const dinoStaff = object.dino;
  console.warn(dinoStaff);
  
  const assignmentArray = [];
  dinoStaff.forEach((element) => {
    const staffKey = element.staff_firebaseKey;
    const dinoKey = element.dino_firebaseKey;
    const promiseOne = getSingleDino(dinoKey);
    const promiseTwo = getSingleStaffMember(staffKey);

    assignmentArray.push(Promise.all([promiseOne, promiseTwo])
        .then(([dinoResponse, StaffResponse]) => ({ dino: dinoResponse.species, staff: StaffResponse.name })));
  });
 
  Promise.all(assignmentArray)
     .then(resolve) // This will resolve whatever the response is. In this case an array of objects
     .catch((error) => reject(error));
});
```


# Option 2: More Sexy
Try something like this: The only difference here is that we don't have to create an empty array because `.map()` does that for us.
```javascript
const getAssignmentStaff = (object) => new Promise((resolve, reject) => {
  const dinoStaff = object.dino;
  const assignmentArray = dinoStaff.map((element) => {
    const staffKey = element.staff_firebaseKey;
    const dinoKey = element.dino_firebaseKey;
    const promiseOne = getSingleDino(dinoKey);
    const promiseTwo = getSingleStaffMember(staffKey);

    return Promise.all([promiseOne, promiseTwo])
       .then(([dinoResponse, StaffResponse]) => ({ dino: dinoResponse.species, staff: StaffResponse.name }));
  });
   Promise.all(assignmentArray)
     .then(resolve) // This will resolve whatever the response is. In this case an array of objects
     .catch((error) => reject(error));
});
```

# Option 3: EVEN More Sexy
Try something like this using Async/Await...AGAIN, I have not run this...but we can talk about it later
```javascript
async function getAssignmentStaff(object) {
  const dinoStaff = object.dino;
  const assignmentArray = dinoStaff.map((element) => {
    const staffKey = element.staff_firebaseKey;
    const dinoKey = element.dino_firebaseKey;
    await promiseOne = getSingleDino(dinoKey);
    await promiseTwo = getSingleStaffMember(staffKey);

    return { dino: dinoResponse.species, staff: StaffResponse.name };
  });
   return assignmentArray
});
```

**Comment by caseywalker** [comment, marked as answer]
# Solution:
None of the above. Let me start by saying: Thank you @drteresavasquez  and @TrinityChristiana. All 3 of these are viable solutions, and a unique way of approaching the way I had approached my promises. However, I found a more efficient way with the video and repo you suggested with the Complex Data structures. 

## Reason for starting from scratch:
I was using two promises to pass all equipment to the promise which was ultimately returning the object or array of objects. This is too complex, and requiring too many .then chains on DOM events, and I noticed that the data was not always pairing the staff to the appropriate assets. 🍝 

## The fix: 
Reduce Two Promises into One Promise. Get all Assets, Get all Staff, Get all Intermediate Data. Wrap all 🎁  that up in a Promise.all, then map through all assets, into an array with only intermediate data with the same asset id, then map through that array and find the staff member's id. Finally, return an object with that asset (spread syntax), and the array of staff members. 

### If that didn't make sense, see code example below. 

``` const dinosWithStaff = () => new Promise((resolve, reject) => {
  Promise.all([getDinos(), getStaff(), getDinoStaff()])
    .then(([dinos, staff, dinosStaffJoin]) => {
      const allStaffInfoArray = dinos.map((dino) => {
        const dinoRelationshipsArray = dinosStaffJoin.filter((dinoStaff) => dinoStaff.dino_firebaseKey === dino.firebaseKey);

        const staffInfoArray = dinoRelationshipsArray.map((staffRelationship) => staff.find((staffMember) => staffMember.firebaseKey === staffRelationship.staff_firebaseKey));

        return { ...dino, staffInfoArray };
      });
      resolve(allStaffInfoArray);
    }).catch((error) => reject(error));
});

**Reply by drteresavasquez** [reply]
Love that the video helped! I thought it would! Great explanation!
